### PR TITLE
Fix some of the issues spotted by SonarCload

### LINKED ISSUE #3292: TypeError in "SummaryInfo._get_column_info_content"
Hello,

While analyzing biopython on SonarCloud I saw what looks like an error in [`Bio/Align/AlignInfo.py`](https://github.com/biopython/biopython/blob/05107f5cfd1f8d67dbc4d472268618f2b6d93bc2/Bio/Align/AlignInfo.py#L490):

Screenshot 2020-10-02 at 18 12 39

You can see [the issue in SonarCloud here](https://sonarcloud.io/project/issues?id=nicolas-harraudeau-sonarsource_biopython&issues=AXTqE7oaA4gH5cbKom8q&open=AXTqE7oaA4gH5cbKom8q).

I don't know what the developer expected to do, but the `-` operator cannot be called on two lists. If this code is reached it will raise the following exception:
```
TypeError: unsupported operand type(s) for -: 'list' and 'list'
```

In case you want to scan biopython on SonarCloud [you can use a configuration similar to this one](https://github.com/nicolas-harraudeau-sonarsource/biopython/commit/9952e28e9c411a626620a78993fd9af8d24dd9ab) when you enable automatic analysis. SonarCloud is free for open source projects.

### PR DESCRIPTION
**Pull request by peterjc**
- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

Closes #3292 and #3294 reported by @nicolas-harraudeau-sonarsource

### CODE REVIEW

**Comment by nicolas-harraudeau-sonarsource** [comment]
Hi @peterjc,

I just saw that you are fixing some other bugs reported by SonarCloud which I didn't have time to report.

In case you consider that some reported bugs are code smells or worse, False Positives, I'd be happy to have your opinion. It will help me redesign the rules detecting these issues. We have [a community forum just for that](https://community.sonarsource.com/c/bug/fp/7).
For example I have some doubts about issues [like this one](https://sonarcloud.io/project/issues?id=nicolas-harraudeau-sonarsource_biopython&issues=AXTqE7c-A4gH5cbKomni&open=AXTqE7c-A4gH5cbKomni). This looks to me like a deliberate choice of design, not a bug. It could be a code smell but I'm not sure yet.

Cheers

**Comment by peterjc** [comment]
@https://github.com/nicolas-harraudeau-sonarsource I've now looked at all the "major" issues in the SonarCloud report. It has found some real issues, and this pull request aims to address a lot of them, and some of the "code smells". Sadly others are things like method and attribute names which we won't change for backward compatibility.

You asked about https://sonarcloud.io/project/issues?id=nicolas-harraudeau-sonarsource_biopython&issues=AXTqE7c-A4gH5cbKomni&open=AXTqE7c-A4gH5cbKomni and I thought it was a mistake in the private API, there was no need to pass the ``line`` variable: 114ff284a7157d8d9bf01b5a042c2d98e4564575

One example of a false positive but nevertheless questionable bit of code is https://sonarcloud.io/project/issues?id=nicolas-harraudeau-sonarsource_biopython&open=AXTqE7izA4gH5cbKomzX&resolved=false&severities=MAJOR&types=BUG (as per the comment line above, deliberate for the side effects). If this was a flake8 warning, I'd handle that with a ``# noqa: ...`` comment on the problematic line.

File: Bio/AlignIO/EmbossIO.py [hunk 1]
```diff
@@ -100,7 +100,7 @@ def __next__(self):
                 % (number_of_seqs, self.records_per_alignment)
             )
 
-        seqs = ["" for id in ids]
+        seqs = ["" for _ in ids]
```
Comment (mdehoon) [inline Bio/AlignIO/EmbossIO.py:103]: or `seqs = [""] * len(ids)`

File: Bio/ExPASy/Prosite.py [hunk 2]
```diff
@@ -300,8 +300,6 @@ def __read(handle):
             record.prorules.extend(rules)
         elif keyword == "DO":
             record.pdoc = value.rstrip(";")
-        elif keyword == "CC":
```
Comment (mdehoon) [inline Bio/ExPASy/Prosite.py:303]: In a way it's nice to keep this to be explicit; i.e. we didn't just forget the `CC` case

File: Bio/Restriction/Restriction.py [hunk 3]
```diff
@@ -202,18 +202,15 @@ def __eq__(self, other):
...
     def linearise(self):
         """Linearise sequence in place."""
         self.linear = True
-        return
 
     def to_linear(self):
         """Make a new instance of sequence as linear."""
         new = self.__class__(self)
         new.linear = True
-        return new
```
Comment (mdehoon) [inline Bio/Restriction/Restriction.py:216]: This one should stay, right?

File: Bio/SearchIO/_legacy/NCBIStandalone.py [hunk 4]
```diff
@@ -1305,7 +1305,7 @@ def align(self, line):
         if len(seq) < self._query_len:
             # Make sure the alignment is the same length as the query
             seq += " " * (self._query_len - len(seq))
-        elif len(seq) < self._query_len:
+        if len(seq) < self._query_len:
```
Comment (mdehoon) [inline Bio/SearchIO/_legacy/NCBIStandalone.py:1308]: Can be part of the previous `if` block

**Review by mdehoon — CHANGES_REQUESTED** [review]
Please see comments

File: Bio/ExPASy/Prosite.py (hunk 2 above)
Comment (peterjc) [reply Bio/ExPASy/Prosite.py:303]: It is already there on line 234, that was the duplication spotted as a likely error.

File: Bio/Restriction/Restriction.py (hunk 3 above)
Comment (peterjc) [reply Bio/Restriction/Restriction.py:216]: Good catch. My mistake.

File: Bio/SearchIO/_legacy/NCBIStandalone.py (hunk 4 above)
Comment (peterjc) [reply Bio/SearchIO/_legacy/NCBIStandalone.py:1308]: True - indented this would be a little clearer.

File: Bio/Restriction/Restriction.py (hunk 3 above)
Comment (peterjc) [reply Bio/Restriction/Restriction.py:216]: The unit tests did catch this:

```
======================================================================
ERROR: test_formatted_seq (test_Restriction.SequenceTesting)
Test several methods of FormattedSeq.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\biopython\Tests\test_Restriction.py", line 69, in test_formatted_seq
    self.assertTrue(linear_seq.is_linear())
AttributeError: 'NoneType' object has no attribute 'is_linear'
======================================================================
```

**Comment by peterjc** [comment]
As to the issue flagged at https://sonarcloud.io/project/issues?id=nicolas-harraudeau-sonarsource_biopython&issues=AXTqE7c-A4gH5cbKomni&open=AXTqE7c-A4gH5cbKomni - I was mislead into thinking the ``line`` variable was not being used in the private functions in ``Bio/motifs/minimal.py`` but it was being updated as a side effect. Should be fixed now by explicitly returning it.

**Comment by peterjc** [comment]
Assuming this passes the tests, I will rebase to make the history a little clearer on ``Bio/motifs/minimal.py`` and ``Bio/Restriction/Restriction.py``, and then we can rebase-and-merge after review.

**Comment by peterjc** [comment]
I have confused myself - tests seem to work locally, will need to come back to the substitution matrix changes when I have more time.

**Comment by peterjc** [comment]
I've broken something to do with alignment substitution matrices:

```
======================================================================
ERROR: test_doctests (test_Tutorial.TutorialTestCase)
Run tutorial doctests.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_Tutorial.py", line 245, in test_doctests
    "%i Tutorial doctests failed: %s" % (len(failures), ", ".join(failures))
ValueError: 6 Tutorial doctests failed: test_chapter_align_line_00864, test_chapter_align_line_01832, test_chapter_align_line_02632, test_chapter_align_line_02746, test_chapter_align_line_02941, test_chapter_cookbook_line_01628
======================================================================
FAIL: test_basics_matrix (test_align_substitution_matrices.Test_basics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_align_substitution_matrices.py", line 76, in test_basics_matrix
    """,
AssertionError: '    X   Y   Z\nX   0.0   0.0   0.0\nY   0.0   0.0   0.0\nZ   0.0   0.0   0.0\n' != '    X   Y   Z\nX 0.0 0.0 0.0\nY 0.0 0.0 0.0\nZ 0.0 0.0 0.0\n'
      X   Y   Z
- X   0.0   0.0   0.0
?  --     --   --
+ X 0.0 0.0 0.0
- Y   0.0   0.0   0.0
?  --     --   --
+ Y 0.0 0.0 0.0
- Z   0.0   0.0   0.0
?  --     --   --
+ Z 0.0 0.0 0.0
======================================================================
FAIL: substitutions (Bio.Align.MultipleSeqAlignment)
Doctest: Bio.Align.MultipleSeqAlignment.substitutions
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/python/3.6.7/lib/python3.6/doctest.py", line 2199, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for Bio.Align.MultipleSeqAlignment.substitutions
  File "/home/travis/build/biopython/biopython/build/lib.linux-x86_64-3.6/Bio/Align/__init__.py", line unknown line number, in substitutions
----------------------------------------------------------------------
File "/home/travis/build/biopython/biopython/build/lib.linux-x86_64-3.6/Bio/Align/__init__.py", line ?, in Bio.Align.MultipleSeqAlignment.substitutions
Failed example:
    print(m)
Expected:
        A   C   G   T
    A 3.0 0.5 0.0 2.5
    C 0.5 1.0 0.0 2.0
    G 0.0 0.0 1.0 1.0
    T 2.5 2.0 1.0 1.0
    <BLANKLINE>
Got:
        A   C   G   T
    A    2.5    2.0    1.0    1.0
    C    2.5    2.0    1.0    1.0
    G    2.5    2.0    1.0    1.0
    T    2.5    2.0    1.0    1.0
    <BLANKLINE>
----------------------------------------------------------------------
```

**Review by mdehoon — APPROVED** [review]

**Comment by peterjc** [comment]
Thank you!

**Pull request merged**